### PR TITLE
add observableProperty decorator

### PR DESCRIPTION
**Pull request by MIYankov**
We can also have one more decorator without the notifyPropertyChanged and clean up the view models further.

### CODE REVIEW

**Comment by manoldonev** [comment]
ObservableProperty.ts file should start with lower-case by our convention.

File: shared/ObservableProperty.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
+import { Observable } from "data/observable";
+
+export function ObservableProperty() {
+    return (target: Observable, propertyKey: string) => {
+        Object.defineProperty(target, propertyKey, {
+            get: function () {
```
Comment (manoldonev) [inline shared/ObservableProperty.ts]: tslint errors here for not using shorthand notation.

File: shared/ObservableProperty.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
+import { Observable } from "data/observable";
+
+export function ObservableProperty() {
+    return (target: Observable, propertyKey: string) => {
+        Object.defineProperty(target, propertyKey, {
+            get: function () {
+                return this["_" + propertyKey];
+            },
+            set: function (value) {
```
Comment (manoldonev) [inline shared/ObservableProperty.ts]: tslint errors here for not using shorthand notation.

File: shared/ObservableProperty.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+            get: function () {
+                return this["_" + propertyKey];
+            },
+            set: function (value) {
+                if (this["_" + propertyKey] === value) {
+                    return;
+                }
+
+                this["_" + propertyKey] = value;
+                this.notify({
```
Comment (manoldonev) [inline shared/ObservableProperty.ts]: You can use the `notifyPropertyChange(propertyName: string, value: any, oldValue?: any): void;` method instead -- internally it does the same but its signature is simpler: `this.notifyPropertyChange(propertyKey, value);`

Btw, I was wondering should we provide the optional oldValue parameter as well?

File: cars/cars-list-view-model.ts [hunk 4]
```diff
@@ -2,15 +2,17 @@ import { Observable } from "data/observable";
...
 import { Car } from "./shared/car-model";
 import { CarService } from "./shared/car-service";
 
 /* ***********************************************************
 * This is the master list view model.
 *************************************************************/
 export class CarsListViewModel extends Observable {
-    private _isLoading: boolean = false;
-    private _cars: ObservableArray<Car> = new ObservableArray<Car>([]);
+    @ObservableProperty() cars: ObservableArray<Car> = new ObservableArray<Car>([]);
```
Comment (manoldonev) [inline cars/cars-list-view-model.ts]: That looks a bit odd to me -- it might be working correctly, just odd.

If we will be using the decorator, should we keep the initialization as is or move it in the constructor?

**Comment by MIYankov** [comment]
Regarding convention - we agreed the proper name should be observable-property-decorator.

Regarding additional decorator - not needed at the moment.

File: cars/cars-list-view-model.ts (hunk 4 above)
Comment (MIYankov) [inline cars/cars-list-view-model.ts]: This should be done in another pull request since there are general differences in view model initialization.

**Comment by MIYankov** [comment]
@manoldonev Check out the applied fixes.

File: shared/observable-property-decorator.ts [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+                }
+
+                this["_" + propertyKey] = value;
+                this.notifyPropertyChange(propertyKey, value);
+            },
+            enumerable: true,
+            configurable: true
+        });
+    };
+}
```
Comment (manoldonev) [inline shared/observable-property-decorator.ts]: file: 'file:///d%3A/GitHub/md-myankov-ts/app/shared/observable-property-decorator.ts'
severity: 'Error'
message: 'file should end with a newline (eofline)'
at: '21,2'
source: 'tslint'
